File: ProjectCS/UI/MenuActionPicker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SessionResultsCS.UI
{
    public class MenuActionPicker
    {
        private QuickItemPicker _picker;
        private IList<Action> _actions;

        public MenuActionPicker(IEnumerable<string> items, IEnumerable<Action> actions)
        {
            _picker = new QuickItemPicker(items);
            _actions = actions.ToList();
            if (_picker.ItemCount != _actions.Count)
            {
                throw new ArgumentException("Item count is not equal to actions count");
            }
        }

        public MenuActionPicker(IEnumerable<(string, Action)> items)
            : this(items.Select((tuple) => tuple.Item1), items.Select((tuple) => tuple.Item2)) { }


        public void PickAndRun()
        {
            int selectedIndex = _picker.Pick();
            _actions[selectedIndex].Invoke();
        }
    }
}

File: Project/UI/MenuActionPicker.cpp
#include "UI/MenuActionPicker.hpp"

#include <stdexcept>


MenuActionPicker::MenuActionPicker(const std::vector<std::string>& items, const std::vector<std::function<void()>>& actions) {
	m_picker = std::make_unique<QuickItemPicker>(items);
	m_actions = actions;
	if (m_picker->GetItemCount() != m_actions.size()) {
		throw std::invalid_argument("Item count is not equal to actions count");
	}
}

MenuActionPicker::MenuActionPicker(const std::vector<Item>& items) {
	std::vector<std::string> labels;
	std::vector<std::function<void()>> actions;

	for (auto p : items) {
		labels.push_back(p.first);
		actions.push_back(p.second);
	}

	m_picker = std::make_unique<QuickItemPicker>(labels);
	m_actions = actions;
}


void MenuActionPicker::PickAndRun() {
	int selectedIndex = m_picker->Pick();
	m_actions[selectedIndex]();
}
